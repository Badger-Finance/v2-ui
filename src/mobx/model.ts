import firebase from 'firebase';
import BigNumber from 'bignumber.js';
import { AbiItem } from 'web3-utils';
import Web3 from 'web3';
import { CustomNotificationObject, EmitterListener, TransactionData } from 'bnc-notify';
import { LockAndMintParamsSimple, BurnAndReleaseParamsSimple } from '@renproject/interfaces';

import { RootStore } from './store';
import { getAirdrops } from 'config/system/airdrops';
import { getGeysers } from '../config/system/geysers';
import { getRebase } from '../config/system/rebase';
import { getRewards } from 'config/system/rewards';
import { NETWORK_IDS, NETWORK_LIST, ZERO, TEN } from 'config/constants';
import { getTokens } from '../config/system/tokens';
import { getVaults } from '../config/system/vaults';
import { getStrategies } from '../config/system/strategies';
import { getNetworkDeploy } from './utils/network';

export class Contract {
	store!: RootStore;
	public address!: string;

	constructor(store: RootStore, address: string) {
		this.store = store;
		this.address = address;
	}
}
export class Token extends Contract {
	public balance!: BigNumber;
	public decimals!: number;
	public totalSupply!: BigNumber;
	public symbol!: string;
	public name!: string;
	public ethValue!: BigNumber;
	public vaults!: Vault[];

	constructor(store: RootStore, address: string, decimals: number) {
		super(store, address);
		this.balance = new BigNumber(0);
		this.ethValue = new BigNumber(0);
		this.decimals = decimals;
		this.vaults = [];
	}

	balanceValue(): BigNumber {
		return this.balance.dividedBy(10 ** this.decimals).multipliedBy(this.ethValue);
	}

	update(payload: TokenPayload): void {
		if (!!payload.balanceOf) this.balance = payload.balanceOf;
		if (!!payload.decimals) this.decimals = payload.decimals;
		if (!!payload.symbol) this.symbol = payload.symbol;
		if (!!payload.ethValue) this.ethValue = payload.ethValue;
		if (!!payload.totalSupply) this.totalSupply = payload.totalSupply;
		if (!!payload.name) this.name = payload.name;
	}
}

export class Vault extends Token {
	public position!: number;
	public holdings!: BigNumber;
	public underlyingToken!: Token;
	public growth!: Growth[];
	public geyser!: Geyser;
	public pricePerShare!: BigNumber;
	public abi!: AbiItem;
	public super!: boolean;
	public vaultBalance!: BigNumber;
	public withdrawAll!: boolean;

	constructor(store: RootStore, address: string, decimals: number, underlyingToken: Token, abi: AbiItem) {
		super(store, address, decimals);
		this.pricePerShare = new BigNumber(1);
		this.underlyingToken = underlyingToken;
		this.underlyingToken.vaults.push(this);
		this.decimals = 18;
		this.holdings = new BigNumber(0);
		this.vaultBalance = new BigNumber(0);
		this.abi = abi;
		this.super = false;
		this.withdrawAll = true;
	}

	deposit(amount: BigNumber): void {
		this.store.contracts.deposit(this, amount);
	}

	withdraw(amount: BigNumber): void {
		this.store.contracts.withdraw(this, amount);
	}

	holdingsValue(): BigNumber {
		return this.holdings
			.multipliedBy(this.pricePerShare)
			.dividedBy(10 ** this.decimals)
			.multipliedBy(this.underlyingToken.ethValue);
	}

	balanceValue(): BigNumber {
		return this.balance
			.multipliedBy(this.pricePerShare)
			.dividedBy(10 ** this.decimals)
			.multipliedBy(this.underlyingToken.ethValue);
	}

	update(payload: TokenPayload): void {
		super.update(payload);
		if (!!payload.position) this.position = payload.position;
		if (!!payload.growth) this.growth = payload.growth;
		if (!!payload.balance) this.vaults = payload.balance;
		if (!!payload.getPricePerFullShare) this.pricePerShare = payload.getPricePerFullShare;
		if (!!payload.totalSupply) this.holdings = payload.totalSupply;
		if ('isSuperSett' in payload) this.super = payload.isSuperSett;
		if (!!payload.ethValue) this.ethValue = payload.ethValue;
		if ('withdrawAll' in payload) this.withdrawAll = payload.withdrawAll;
	}
}

export class Geyser extends Contract {
	public vault!: Vault;
	public holdings!: BigNumber;
	public balance!: BigNumber;
	public rewards!: Growth[];
	public abi!: AbiItem;

	constructor(store: RootStore, address: string, vault: Vault, abi: AbiItem) {
		super(store, address);
		this.vault = vault;
		this.vault.geyser = this;
		this.holdings = new BigNumber(0);
		this.balance = new BigNumber(0);
		this.abi = abi;
	}

	stake(amount: BigNumber): void {
		this.store.contracts.stake(this.vault, amount);
	}

	unstake(amount: BigNumber): void {
		this.store.contracts.unstake(this.vault, amount);
	}

	holdingsValue(): BigNumber {
		return this.holdings
			.dividedBy(10 ** this.vault.decimals)
			.multipliedBy(this.vault.pricePerShare)
			.multipliedBy(this.vault.underlyingToken.ethValue);
	}

	balanceValue(): BigNumber {
		return this.balance
			.dividedBy(10 ** this.vault.decimals)
			.multipliedBy(this.vault.pricePerShare)
			.multipliedBy(this.vault.underlyingToken.ethValue);
	}

	update(payload: GeyserPayload): void {
		if (!!payload.totalStaked) this.holdings = payload.totalStaked;
		if (!!payload.totalStakedFor) this.balance = payload.totalStakedFor;
	}
}

export interface StrategyConfig {
	name: string;
	address: string;
	fees: FeeConfig;
	strategyLink: string;
}

export interface StrategyNetworkConfig {
	[vaultAddress: string]: StrategyConfig;
}

export interface FeeConfig {
	[feeName: string]: BigNumber;
}

export class TokenModel extends Contract {
	public name: string;
	public symbol: string;
	public decimals: number;
	public balance: BigNumber;
	public poolId?: number | undefined;
	public mintRate: string;
	public redeemRate: string;

	constructor(store: RootStore, data: TokenConfig) {
		super(store, Web3.utils.toChecksumAddress(data.address));
		this.name = data.name;
		this.symbol = data.symbol;
		this.decimals = data.decimals;
		this.poolId = data?.poolId;
		this.balance = ZERO;
		// This will be fetched and set at initialization using 1 unit of mint and redeem
		// to show current conversion rate from token to ibBTC and from ibBTC to token
		// by fetchConversionRates()
		this.mintRate = '0.000';
		this.redeemRate = '0.000';
	}

	public get formattedBalance(): string {
		return this.unscale(this.balance).toFixed(3);
	}

	public get icon(): any {
		return `/assets/icons/${this.symbol.toLowerCase()}.svg`;
	}

	public formatAmount(amount: BigNumber | string): string {
		return this.unscale(new BigNumber(amount)).toFixed(3);
	}

	public scale(amount: BigNumber | string): BigNumber {
		return new BigNumber(amount).multipliedBy(TEN.pow(this.decimals));
	}

	public unscale(amount: BigNumber | string): BigNumber {
		return new BigNumber(amount).dividedBy(TEN.pow(this.decimals));
	}
}

export interface BadgerTree {
	cycle: string;
	timeSinceLastCycle: string;
	sharesPerFragment: BigNumber | undefined;
	proof: RewardMerkleClaim | undefined;
	claims: UserClaimData[] | undefined;
	claimableAmounts: BigNumber[] | undefined;
}

interface TokenConfig {
	address: string;
	name: string;
	symbol: string;
	decimals: number;
	poolId?: number | undefined;
	mintRate?: BigNumber | string;
	redeemRate?: string;
}

export interface Growth {
	day: Amount;
	week: Amount;
	month: Amount;
	year: Amount;
}

export interface RewardMerkleClaim {
	index: string;
	cycle: string;
	boost: BigNumber;
	user: string;
	tokens: string[];
	cumulativeAmounts: string[];
	proof: string[];
	node: string;
}

export type TreeClaimData = [string[], BigNumber[]];

export interface UserClaimData {
	token: string;
	amount: BigNumber;
}

export interface Amount {
	token: Token;
	amount: BigNumber;
}

export type ReduceAirdropsProps = {
	digg?: BigNumber;
	merkleProof?: any;
	bBadger?: BigNumber;
};

export type TokenRebaseStats = {
	totalSupply: BigNumber;
	decimals: number;
	lastRebaseTimestampSec: number;
	minRebaseTimeIntervalSec: number;
	rebaseLag: any;
	epoch: any;
	inRebaseWindow: boolean;
	rebaseWindowLengthSec: number;
	oracleRate: BigNumber;
	derivedEth: any;
	nextRebase: Date;
	pastRebase: any;
};

export type TokenPayload = {
	balanceOf: BigNumber;
	decimals: number;
	symbol: string;
	ethValue: BigNumber;
	totalSupply: BigNumber;
	name: string;
	position: number;
	growth: Growth[];
	balance: Vault[];
	getPricePerFullShare: BigNumber;
	isSuperSett: boolean;
	withdrawAll: boolean;
};

export type GeyserPayload = {
	totalStaked: BigNumber;
	totalStakedFor: BigNumber;
};

export type Schedules = {
	[index: string]: string[][];
};

export type RebaseToStats = { nextRebase: Date; oracleRate: string; totalSupply: string | boolean };

export type ContractToStats = {
	stats: {
		tvl: BigNumber;
		portfolio: BigNumber;
		wallet: BigNumber;
		bDigg: any;
		deposits: BigNumber;
		badger: BigNumber;
		digg: BigNumber;
		vaultDeposits: BigNumber;
	};
};

export type ReducedAirdops = {
	digg?: {
		amount: BigNumber;
		token: any;
	};
	bBadger?: {
		amount: BigNumber | null;
		token: any;
	};
};

export type MethodConfigPayload = { [index: string]: string };

export type ReducedCurveResult = { address: any; virtualPrice: BigNumber; ethValue: BigNumber }[];

export type TokenAddressessConfig = {
	underlying: any;
	sushi: string[];
};

export type TokenAddressess = {
	address: any;
	contract: string;
	type: string;
	subgraph: string;
};

export type ReducedContractConfig = {
	defaults: Record<any, any>;
	batchCall: {
		namespace: string;
		addresses: string[];
		allReadMethods: boolean;
		groupByNamespace: boolean;
		logging: boolean;
	}[];
};

export type ContractMethodsConfig = {
	rewards: {
		method: any;
		tokens: any;
	};
	walletMethods: string[];
};

export type BatchConfig = {
	namespace: string;
	addresses: string[];
	allReadMethods: boolean;
	groupByNamespace: boolean;
	logging: boolean;
};

export type ReducedGraphResults = {
	address: string;
	ethValue: BigNumber;
	name: string;
	type: string;
};

export type GraphResultPrices = {
	[x: string]: {
		ethValue: BigNumber;
	};
};

export type TokenContract = {
	contract: string;
};

export type AirdropsConfig = {
	[index: string]: {
		tokenAbi: AbiItem[];
		tokenContract: string | { [index: string]: string };
		airdropContract: string;
		airdropAbi: AbiItem[];
	};
};

export type AirdropNetworkConfig = {
	airdropEndpoint: string;
	airdropsConfig: AirdropsConfig;
};

export type VaultNetworkConfig = {
	[index: string]:
		| {
				abi: AbiItem[];
				underlying: string;
				contracts: string[];
				fillers: {
					[index: string]: string[] | boolean[] | number[];
				};
				methods: {
					name: string;
					args?: string[];
				}[];
				growthEndpoints?: string[];
		  }
		| undefined;
};

export type GeyserNetworkConfig = {
	geyserBatches?: {
		abi: AbiItem[];
		underlying: string;
		methods: {
			name: string;
			args?: string[];
		}[];
		contracts: string[];
		fillers?: {
			isFeatured?: boolean[];
			isSuperSett?: boolean[];
			getStakingToken?: string[];
		};
	}[];
};

export type RewardNetworkConfig = {
	endpoint: string;
	network: number;
	contract: string;
	abi: AbiItem[];
	tokens: string[];
};

export type TokenContractInfo = {
	abi: AbiItem[];
	methods: {
		name: string;
		args?: string[];
	}[];
	contracts: string[];
};

export type TokenNetworkConfig = {
	curveTokens?: {
		contracts: string[];
		priceEndpoints: string[];
		names: string[];
		vsToken: string;
	};
	priceEndpoints: string[];
	tokenBatches: TokenContractInfo[];
	decimals: { [index: string]: number };
	symbols: { [index: string]: string };
	names: { [index: string]: string };
	tokenMap: { [index: string]: string };
};

export type RebaseNetworkConfig = {
	digg: {
		addresses: string[];
		abi: AbiItem[];
		allReadMethods?: boolean;
		readMethods?: {
			name: string;
			args: (string | number)[];
		}[];
		groupByNamespace: boolean;
		logging?: boolean;
		namespace: string;
	}[];
	orchestrator: {
		contract: string;
		abi: AbiItem[];
	};
};

export type BadgerSystem = {
	[index: string]: string | { [index: string]: string };
};

export type DeployConfig = {
	[index: string]: any;
};

export type NetworkConstants = {
	[index: string]: {
		APP_URL: string;
		RPC_URL: string;
		START_BLOCK: number;
		START_TIME: Date;
		DEPLOY: DeployConfig;
	};
};

export type ClaimsSymbols = {
	[index: string]: {
		[index: string]: string;
	};
};

export interface GasPrices {
	[speed: string]: number;
}

export interface NotifyLink extends CustomNotificationObject {
	link: string;
}

export interface Network {
	name: string;
	networkId: number;
	fullName: string;
	tokens: TokenNetworkConfig;
	vaults: VaultNetworkConfig;
	geysers: GeyserNetworkConfig;
	rebase: RebaseNetworkConfig | undefined;
	airdrops: AirdropNetworkConfig | undefined;
	deploy: DeployConfig;
	rewards: RewardNetworkConfig | undefined;
	currency: string;
	gasEndpoint: string;
	sidebarTokenLinks: {
		url: string;
		title: string;
	}[];
	settOrder: string[];
	getGasPrices: () => Promise<GasPrices>;
	getNotifyLink: EmitterListener;
	isWhitelisted: { [index: string]: boolean };
	cappedDeposit: { [index: string]: boolean };
	uncappedDeposit: { [index: string]: boolean };
	newVaults: { [index: string]: string[] };
	strategies: StrategyNetworkConfig;
	getFees: (vaultAddress: string) => string[];
}

const _getFees = (strategy: StrategyConfig) => {
	const feeList: string[] = [];
	if (!strategy) return [];
	// fees are stored in BIPs on the contract, dividing by 10**2 makes them readable %
	Object.keys(strategy.fees).forEach((key) => {
		const value = strategy.fees[key];
		if (value.gt(0)) {
			feeList.push(`${key}:  ${value.dividedBy(10 ** 2).toString()}%`);
		}
	});
	return feeList;
};

export class BscNetwork implements Network {
	public readonly name = NETWORK_LIST.BSC;
	public readonly networkId = NETWORK_IDS.BSC;
	public readonly fullName = 'Binance Smart Chain';
	public readonly tokens = getTokens(NETWORK_LIST.BSC);
	public readonly vaults = getVaults(NETWORK_LIST.BSC);
	public readonly geysers = getGeysers(NETWORK_LIST.BSC);
	public readonly rebase = getRebase(NETWORK_LIST.BSC);
	public readonly airdrops = getAirdrops(NETWORK_LIST.BSC);
	public readonly deploy = getNetworkDeploy(NETWORK_LIST.BSC);
	public readonly rewards = getRewards(NETWORK_LIST.BSC);
	public readonly currency = 'BNB';
	public readonly gasEndpoint = '';
	// Deterministic order for displaying setts on the sett list component
	public readonly settOrder = [
		this.deploy.sett_system.vaults['native.bDiggBtcb'],
		this.deploy.sett_system.vaults['native.bBadgerBtcb'],
		this.deploy.sett_system.vaults['native.pancakeBnbBtcb'],
		this.deploy.sett_system.vaults['yearn.wBtc'],
	];
	public readonly sidebarTokenLinks = [
		{
			url: 'https://pancakeswap.info/pair/0xE1E33459505bB3763843a426F7Fd9933418184ae',
			title: 'PancakeSwap bDigg/BtcB',
		},
		{
			url: 'https://pancakeswap.info/pair/0x10f461ceac7a17f59e249954db0784d42eff5db5',
			title: 'PancakeSwap bBadger/BtcB',
		},
	];
	public async getGasPrices(): Promise<GasPrices> {
		return { standard: 5 };
	}
	public getNotifyLink(transaction: TransactionData): NotifyLink {
		return { link: `https://bscscan.com//tx/${transaction.hash}` };
	}
	public readonly isWhitelisted = {};
	public readonly cappedDeposit = {};
	public readonly uncappedDeposit = {};
	public readonly newVaults = {};
	public readonly strategies = getStrategies(NETWORK_LIST.BSC);
	public getFees(vaultAddress: string): string[] {
		return _getFees(this.strategies[vaultAddress]);
	}
}

export class EthNetwork implements Network {
	public readonly name = NETWORK_LIST.ETH;
	public readonly networkId = NETWORK_IDS.ETH;
	public readonly fullName = 'Ethereum';
	public readonly tokens = getTokens(NETWORK_LIST.ETH);
	public readonly vaults = getVaults(NETWORK_LIST.ETH);
	public readonly geysers = getGeysers(NETWORK_LIST.ETH);
	public readonly rebase = getRebase(NETWORK_LIST.ETH);
	public readonly airdrops = getAirdrops(NETWORK_LIST.ETH);
	public readonly deploy = getNetworkDeploy(NETWORK_LIST.ETH);
	public readonly rewards = getRewards(NETWORK_LIST.ETH);
	public readonly currency = 'ETH';
	public readonly gasEndpoint = 'https://www.gasnow.org/api/v3/gas/price?utm_source=badgerv2';
	// Deterministic order for displaying setts on the sett list component
	public readonly settOrder = [
		this.deploy.sett_system.vaults['native.sushiibBTCwBTC'],
		this.deploy.sett_system.vaults['yearn.wBtc'],
		this.deploy.sett_system.vaults['native.digg'],
		this.deploy.sett_system.vaults['native.badger'],
		this.deploy.sett_system.vaults['native.sushiDiggWbtc'],
		this.deploy.sett_system.vaults['native.sushiBadgerWbtc'],
		this.deploy.sett_system.vaults['native.sushiWbtcEth'],
		this.deploy.sett_system.vaults['native.uniDiggWbtc'],
		this.deploy.sett_system.vaults['native.uniBadgerWbtc'],
		this.deploy.sett_system.vaults['native.renCrv'],
		this.deploy.sett_system.vaults['native.sbtcCrv'],
		this.deploy.sett_system.vaults['native.tbtcCrv'],
		this.deploy.sett_system.vaults['harvest.renCrv'],
	];
	public readonly sidebarTokenLinks = [
		{
			url: 'https://matcha.xyz/markets/BADGER',
			title: 'BADGER',
		},
		{
			url: 'https://info.uniswap.org/pair/0xcd7989894bc033581532d2cd88da5db0a4b12859',
			title: 'Uniswap BADGER/wBTC',
		},
		{
			url: 'https://analytics.sushi.com/pairs/0x110492b31c59716ac47337e616804e3e3adc0b4a',
			title: 'Sushiswap BADGER/wBTC',
		},
	];
	public async getGasPrices(): Promise<GasPrices> {
		const prices = await fetch('https://www.gasnow.org/api/v3/gas/price?utm_source=badgerv2');
		const result = await prices.json();
		return {
			rapid: result.data['rapid'] / 1e9,
			fast: result.data['fast'] / 1e9,
			standard: result.data['standard'] / 1e9,
			slow: result.data['slow'] / 1e9,
		};
	}
	public getNotifyLink(transaction: TransactionData): NotifyLink {
		return { link: `https://etherscan.io/tx/${transaction.hash}` };
	}
	public readonly isWhitelisted = {};
	public readonly cappedDeposit = {};
	public readonly uncappedDeposit = {
		[this.deploy.sett_system.vaults['yearn.wBtc']]: true,
	};
	public readonly newVaults = {
		[this.deploy.sett_system.vaults['yearn.wBtc']]: ['Expected ROI', '60% @ $100m', '30% @ $400m', '24% @ $1b'],
		[this.deploy.sett_system.vaults['native.sushiibBTCwBTC']]: [
			'New vault ROIs are determined once there is enough historical data',
		],
	};
	public readonly strategies = getStrategies(NETWORK_LIST.ETH);
	public getFees(vaultAddress: string): string[] {
		return _getFees(this.strategies[vaultAddress]);
	}
}

export type UserPermissions = {
	viewSettShop: boolean;
};

export type Eligibility = {
	isEligible: boolean;
};

export type BouncerProof = {
	address: string;
	proof: string[];
};

export interface BoostMultipliers {
	[contract: string]: number;
}

export interface Account {
	id: string;
	boost: number;
	multipliers: BoostMultipliers;
	depositLimits: AccountLimits;
	// currently unused below
	value: number;
	earnedValue: number;
	balances: SettBalance[];
}

export interface SettBalance {
	id: string;
	name: string;
	asset: string;
	balance: TokenBalance[];
	value: number;
	earnedTokens: TokenBalance[];
	earnedValue: number;
}

export interface AccountLimits {
	[contract: string]: DepositLimit;
}

export interface DepositLimit {
	available: number;
	limit: number;
}

export enum Protocol {
	Curve = 'curve',
	Sushiswap = 'sushiswap',
	Uniswap = 'uniswap',
	Pancakeswap = 'pancakeswap',
	Yearn = 'yearn',
	Harvest = 'harvest',
}

/**
 * Sett and geyser objects will be represented by the same
 * interface. The key difference between a sett and geyser
 * is the value sources which populate the entity. Geyser will
 * have emissions value sources while setts only have the
 * native underlying value source.
 */
export interface Sett extends SettSummary {
	asset: string;
	apy: number;
	apr: number;
	minApr?: number;
	maxApr?: number;
	boostable: boolean;
	hasBouncer: boolean;
	ppfs: number;
	sources: ValueSource[];
	tokens: TokenBalance[];
	underlyingToken: string;
	vaultToken: string;
	affiliate?: SettAffiliateData;
	experimental: boolean;
}

export interface SettAffiliateData {
	availableDepositLimit?: number;
	protocol: Protocol;
	depositLimit?: number;
}

export type ValueSource = {
	name: string;
	apy: number;
	apr: number;
	performance: Performance;
	boostable: boolean;
	harvestable: boolean;
	minApr: number;
	maxApr: number;
};

export type Performance = {
	oneDay?: number;
	threeDay?: number;
	sevenDay?: number;
	thirtyDay?: number;
};

export type TokenBalance = {
	address: string;
	name: string;
	symbol: string;
	decimals: number;
	balance: number;
	value: number;
};

export type PriceSummary = {
	[address: string]: BigNumber | undefined;
};

export interface SettSummary {
	name: string;
	value: number;
	balance: number;
}

export type ProtocolSummary = {
	totalValue: number;
	setts?: SettSummary[];
};

export type SettMap = { [contract: string]: Sett };

export type RenVMTransaction = {
	// ID is the pkey in the db.
	id: string;
	user: string;
	// Nonce monotonically increases per user tx.
	nonce: number;
	encodedTx: string; // json encoded tx data.
	// NB: The web3Provider field is not encoded (for obvious reasons).
	params: LockAndMintParamsSimple | BurnAndReleaseParamsSimple;
	status: string;
	// Record if there was an error processing a tx.
	error: string;
	updated: firebase.firestore.Timestamp;
	created: firebase.firestore.Timestamp;
	deleted: boolean;
};

export interface NFT {
	tokenId: string;
	balance: string;
	poolBalance: string;
	totalSupply: string;
	root: string;
	name?: string;
	image?: string;
	redirectUrl?: string;
}

export interface ExchangeRates {
	usd: number;
	cad: number;
	btc: number;
	bnb: number;
}

export interface ibBTCFees {
	mintFeePercent: BigNumber;
	redeemFeePercent: BigNumber;
}
