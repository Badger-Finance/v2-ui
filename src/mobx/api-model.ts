/**
 * All API Model Objects should be directly in line with API interface model
 * objects to ensure data compatiblity. Interfaces located at link below:
 * https://github.com/axejintao/badger-api/tree/staging/src/interface
 */
import BigNumber from "bignumber.js";

export interface Geyser {
  emissions: Emission[],
}

export interface Token {
  address: string;
  name: string;
  symbol: string;
  decimals: number;
}

export interface Emission {
  token: Token,
  unlockSchedule: UnlockSchedule,
}

export interface UnlockSchedule {
	initialLocked: BigNumber,
	endAtSec: BigNumber,
	durationSec: BigNumber,
	startTime: BigNumber,
}

export interface TokenBalance {
  address: string;
  name: string;
  symbol: string;
  decimals: number;
  balance: number,
  value: number,
}

export interface ValueSource {
  name: string,
  apy: number,
  performance: Performance,
}

export interface Sett {
  name: string,
  asset: string,
  ppfs: number,
  value: number,
  apy: number,
  tokens: TokenBalance[],
  sources: ValueSource[],
  geyser?: Geyser,
}

export interface SettSummary {
  name: string,
  asset: string,
  value: number,
  tokens: TokenBalance[],
}
