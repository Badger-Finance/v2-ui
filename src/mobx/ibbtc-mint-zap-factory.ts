import { IbBTCMintZap } from './model/vaults/ibbtc-mint-zap';
import { RootStore } from './RootStore';
import addresses from 'config/ibBTC/addresses.json';
import { TokenZap } from './model/vaults/token-zap';
import { GeneralVaultZap } from './model/vaults/general-vault-zap';
import { RenVaultZap } from './model/vaults/ren-vault-zap';
import { Token } from '@badger-dao/sdk';

export class IbBTCMintZapFactory {
  private static tokenZaps: Record<string, TokenZap> = {};
  private static generalVaultZaps: Record<string, GeneralVaultZap> = {};
  private static renVaultZaps: Record<string, RenVaultZap> = {};

  static getIbBTCZap(store: RootStore, token: Token): IbBTCMintZap {
    if (addresses.mainnet.contracts.TokenZap.supportedTokens.includes(token.address)) {
      if (!this.tokenZaps[token.address]) {
        this.tokenZaps[token.address] = new TokenZap(store, token);
      }
      return this.tokenZaps[token.address];
    }

    if (addresses.mainnet.contracts.GeneralVaultZap.supportedTokens.includes(token.address)) {
      if (!this.generalVaultZaps[token.address]) {
        this.generalVaultZaps[token.address] = new GeneralVaultZap(store, token);
      }
      return this.generalVaultZaps[token.address];
    }

    if (addresses.mainnet.contracts.RenVaultZap.supportedTokens.includes(token.address)) {
      if (!this.renVaultZaps[token.address]) {
        this.renVaultZaps[token.address] = new RenVaultZap(store, token);
      }
      return this.renVaultZaps[token.address];
    }

    throw new Error(`Unrecognized peak type for token => ${token.symbol}`);
  }
}
