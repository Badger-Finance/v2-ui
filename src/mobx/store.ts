import { RouterStore } from 'mobx-router';
import UiState from './reducers';
import WalletStore from './stores/walletStore';
import ContractsStore from './stores/contractsStore';
import AirdropStore from './stores/airdropStore';
import RebaseStore from './stores/rebaseStore';
import RewardsStore from './stores/rewardsStore';
import IbBTCStore from './stores/ibBTCStore';
import TransactionsStore from './stores/transactionsStore';
import SettStoreV2 from './stores/settStoreV2';
import { NETWORK_LIST } from '../config/constants';
import HoneyPotStore from './stores/honeyPotStore';

export class RootStore {
	public router: RouterStore<RootStore>;
	public wallet: WalletStore;
	public uiState: UiState;
	public contracts: ContractsStore;
	public airdrops: AirdropStore;
	public rebase: RebaseStore;
	public rewards: RewardsStore;
	public ibBTCStore: IbBTCStore;
	public setts: SettStoreV2;
	public transactions: TransactionsStore;
	public honeyPot: HoneyPotStore;

	constructor() {
		this.router = new RouterStore<RootStore>(this);
		this.wallet = new WalletStore(this);
		this.contracts = new ContractsStore(this);
		this.airdrops = new AirdropStore(this);
		this.rebase = new RebaseStore(this);
		this.rewards = new RewardsStore(this);
		this.uiState = new UiState(this);
		this.ibBTCStore = new IbBTCStore(this);
		this.honeyPot = new HoneyPotStore(this);
		// RenVM transactions store.
		this.transactions = new TransactionsStore(this);
		this.setts = new SettStoreV2(this);

		this.walletRefresh();
	}

	async walletRefresh(): Promise<void> {
		const chain = this.wallet.network.name;
		const refreshData = [
			this.setts.loadAssets(chain),
			this.setts.loadPrices(chain),
			this.wallet.getGasPrice(),
			this.contracts.updateProvider(),
		];
		if (chain === NETWORK_LIST.ETH) {
			refreshData.push(this.setts.loadGeysers(chain));
			refreshData.push(this.rebase.fetchRebaseStats());
		} else {
			refreshData.push(this.setts.loadSetts(chain));
		}
		await Promise.all(refreshData);

		if (this.wallet.connectedAddress) {
			this.contracts.updateProvider();
			await this.wallet.getGasPrice();
			await this.contracts.fetchContracts();
			if (chain === NETWORK_LIST.ETH) {
				this.uiState.reduceRebase();
				this.ibBTCStore.init();
				this.rewards.fetchSettRewards();
				this.uiState.reduceTreeRewards();
				this.airdrops.fetchAirdrops();
				this.uiState.reduceAirdrops();
			}
			this.uiState.reduceStats();
		}
	}
}

const store = new RootStore();

export default store;
