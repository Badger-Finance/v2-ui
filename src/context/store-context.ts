import { createContext } from "react";
import { RootStore } from "../mobx/store";

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;