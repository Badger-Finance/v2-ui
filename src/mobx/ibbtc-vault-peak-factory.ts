import { IbbtcVaultPeak } from './model/vaults/ibbtc-vault-peak';
import { RootStore } from './store';
import addresses from 'config/ibBTC/addresses.json';
import { ZapPeak } from './model/vaults/zap-peak';
import { YearnPeak } from './model/vaults/yearn-peak';
import { BadgerPeak } from './model/vaults/badger-peak';
import { IbbtcOptionToken } from './model/tokens/ibbtc-option-token';

export class IbbtcVaultPeakFactory {
	static createIbbtcVaultPeakForToken(store: RootStore, token: IbbtcOptionToken): IbbtcVaultPeak {
		if (addresses.mainnet.contracts.ZapPeak.supportedTokens.includes(token.symbol)) {
			return new ZapPeak(store, token);
		}

		if (addresses.mainnet.contracts.yearnWBTCPeak.supportedTokens.includes(token.symbol)) {
			return new YearnPeak(store, token);
		}

		if (addresses.mainnet.contracts.BadgerSettPeak.supportedTokens.includes(token.symbol)) {
			return new BadgerPeak(store, token);
		}

		throw new Error(`Unrecognized peak type for token => ${token.symbol}`);
	}
}
