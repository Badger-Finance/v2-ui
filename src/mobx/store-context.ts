import { RootStore } from './store';
import { createContext } from 'react';

// Use React context to make your store available in your application
export const StoreContext = createContext({} as RootStore);
export const StoreProvider = StoreContext.Provider;
