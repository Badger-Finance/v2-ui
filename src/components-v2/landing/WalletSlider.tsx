import React, { useContext } from 'react';
import { FormControlLabel, Switch, makeStyles } from '@material-ui/core';
import { observer } from 'mobx-react-lite';
import { StoreContext } from '../../mobx/store-context';

const useStyles = makeStyles(() => ({
	walletSlider: {
		height: 36,
	},
}));

const WalletSlider = observer(() => {
	const classes = useStyles();
	const store = useContext(StoreContext);

	const {
		uiState: { showUserBalances, setShowUserBalances },
	} = store;

	return (
		<FormControlLabel
			control={
				<Switch
					checked={showUserBalances}
					onChange={() => setShowUserBalances(!showUserBalances)}
					color="primary"
				/>
			}
			label="Portfolio View"
			className={classes.walletSlider}
		/>
	);
});

export default WalletSlider;
