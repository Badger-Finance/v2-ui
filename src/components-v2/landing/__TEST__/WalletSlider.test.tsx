import React from 'react';
import WalletSlider from '../WalletSlider';
import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import { StoreProvider } from '../../../mobx/store-context';
import store from '../../../mobx/store';

describe('WalletSlider', () => {
	const testStore = store;
	test('Renders correctly', () => {
		const { container } = render(
			<StoreProvider value={testStore}>
				<WalletSlider />
			</StoreProvider>,
		);
		expect(container).toMatchSnapshot();
	});
	// Clicking on slider produces changes in other part of the DOM, to be tested on integration (Landing.test.tsx)
});
