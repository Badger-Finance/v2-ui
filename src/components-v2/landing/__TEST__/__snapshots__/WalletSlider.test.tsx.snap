// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WalletSlider Renders correctly 1`] = `
<div>
  <label
    class="MuiFormControlLabel-root makeStyles-walletSlider-1"
  >
    <span
      class="MuiSwitch-root"
    >
      <span
        aria-disabled="false"
        class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-2 MuiSwitch-switchBase MuiSwitch-colorPrimary"
      >
        <span
          class="MuiIconButton-label"
        >
          <input
            class="PrivateSwitchBase-input-5 MuiSwitch-input"
            type="checkbox"
            value=""
          />
          <span
            class="MuiSwitch-thumb"
          />
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiSwitch-track"
      />
    </span>
    <span
      class="MuiTypography-root MuiFormControlLabel-label MuiTypography-body1"
    >
      Portfolio View
    </span>
  </label>
</div>
`;
